### PR TITLE
feat(langchain): tag tool call outputs for langchain chat models

### PR DESCRIPTION
**Pull request by sabrenner**
## What does this PR do?
[MLOB-1165]

**Note**: Most LOC are from requirements changes, adding cassettes for langchain-specific formatted tool calls, and snapshots.

Adds tagging for LCEL [tool/function calling](https://python.langchain.com/v0.1/docs/modules/model_io/chat/function_calling/). **Specifically, for tool call output.** The two common LangChain methods associated with this, `bind_tools` and `with_structured_output`, are responsible for making tool calls to the respective model and including the tool call results in the LangChain chat model response.

Tool calls are a typed property of the [`AiMessage` LangChain message type](https://github.com/langchain-ai/langchain/blob/1f01c0fd98d1cb768d950b8e46701bf949865af3/libs/core/langchain_core/messages/ai.py#L72). While tool calls from different providers/models are structured differently (ie Anthropic content blocks vs OpenAI tool calls, [ref](https://python.langchain.com/v0.2/docs/how_to/function_calling/#passing-tool-outputs-to-model) under the intro to this documentation), LangChain coalesces these differences under a common [ToolCall](https://github.com/langchain-ai/langchain/blob/1f01c0fd98d1cb768d950b8e46701bf949865af3/libs/core/langchain_core/messages/tool.py#L127-L154) struct. This is what we tag in this.

## Follow-up
1. **Tagging tool calls for LLM Observability will be a separate PR right after this one** (tests will re-use cassettes but that PR will validate we set the tool calls appropriately).
2. This PR covers the scope for `bind_tools`. However, `with_structured_output` actually appends an `OutputParser` to the chat model invocation to transform those tool calls into structured output for the chain. **Tracing and tagging `OutputParsers` will be a separate PR.**

### OpenAI
Before
Screenshot 2024-07-22 at 3 13 52 PM

After
Screenshot 2024-07-22 at 3 14 17 PM

### Anthropic
Before
Screenshot 2024-07-22 at 3 14 41 PM

After
Screenshot 2024-07-22 at 3 14 58 PM

## Checklist
- [x] PR author has checked that all the criteria below are met
- [x] The PR description includes an overview of the change
- [x] The PR description articulates the motivation for the change
- [x] The change includes tests OR the PR description describes a testing strategy
- [x] The PR description notes risks associated with the change, if any
- [x] Newly-added code is easy to change
- [x] The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- [x] The change includes or references documentation updates if necessary
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Newly-added code is easy to change
- [x] Release note makes sense to a user of the library
- [x] If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

[MLOB-1165]: https://datadoghq.atlassian.net/browse/MLOB-1165?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ddtrace/contrib/langchain/patch.py [hunk 1]
```diff
@@ -397,10 +397,36 @@ def traced_chat_model_generate(langchain, pin, func, instance, args, kwargs):
...
+                            integration.trunc(chat_completion.text),
+                        )
+                    if tool_calls:
+                        if not isinstance(tool_calls, list):
+                            tool_calls = [tool_calls]
+                        for tool_call_idx, tool_call in enumerate(tool_calls):
+                            span.set_tag_str(
+                                "langchain.response.completions.%d.%d.tool_calls.%d.id"
+                                % (message_set_idx, idx, tool_call_idx),
+                                integration.trunc(str(tool_call.get("id", ""))),
```
Comment (Yun-Kim) [inline ddtrace/contrib/langchain/patch.py:417]: In the context of a tool call we only consider the output arguments as the completion, meaning we shouldn't need to truncate the completion/tool ID/name

File: ddtrace/contrib/langchain/patch.py [hunk 2]
```diff
@@ -397,10 +397,36 @@ def traced_chat_model_generate(langchain, pin, func, instance, args, kwargs):
...
+                            )
+                            span.set_tag_str(
+                                "langchain.response.completions.%d.%d.tool_calls.%d.name"
+                                % (message_set_idx, idx, tool_call_idx),
+                                integration.trunc(str(tool_call.get("name", ""))),
```
Comment (Yun-Kim) [inline ddtrace/contrib/langchain/patch.py:422]: Same here

File: ddtrace/contrib/langchain/patch.py [hunk 3]
```diff
@@ -502,10 +528,35 @@ async def traced_chat_model_agenerate(langchain, pin, func, instance, args, kwar
...
+                            integration.trunc(chat_completion.text),
+                        )
+                    if tool_calls:
+                        if not isinstance(tool_calls, list):
+                            tool_calls = [tool_calls]
+                        for tool_call_idx, tool_call in enumerate(tool_calls):
+                            span.set_tag_str(
+                                "langchain.response.completions.%d.%d.tool_calls.%d.id"
+                                % (message_set_idx, idx, tool_call_idx),
+                                integration.trunc(str(tool_call.get("id", ""))),
```
Comment (Yun-Kim) [inline ddtrace/contrib/langchain/patch.py:547]: +1

File: ddtrace/contrib/langchain/patch.py [hunk 4]
```diff
@@ -502,10 +528,35 @@ async def traced_chat_model_agenerate(langchain, pin, func, instance, args, kwar
...
+                            )
+                            span.set_tag_str(
+                                "langchain.response.completions.%d.%d.tool_calls.%d.name"
+                                % (message_set_idx, idx, tool_call_idx),
+                                integration.trunc(str(tool_call.get("name", ""))),
```
Comment (Yun-Kim) [inline ddtrace/contrib/langchain/patch.py:552]: +1

File: tests/contrib/langchain/test_langchain_community.py [hunk 5]
```diff
@@ -1267,7 +1267,46 @@ def test_lcel_chain_non_dict_input(langchain_core):
     sequence.invoke(1)
 
 
-@pytest.mark.snapshot
+@pytest.mark.snapshot(ignores=["metrics.langchain.tokens.total_cost"])
```
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_community.py:1270]: For the sake of our flaky langchain community tests, let's add a `@flaky` marker for the new tests while we figure out how to resolve this 😄

File: tests/contrib/langchain/test_langchain_community.py [hunk 6]
```diff
@@ -1267,7 +1267,46 @@ def test_lcel_chain_non_dict_input(langchain_core):
...
+        """
+        return a + b
+
+    llm = langchain_openai.ChatOpenAI(model="gpt-3.5-turbo-0125")
+    llm_with_tools = llm.bind_tools([add])
+    with request_vcr.use_cassette("lcel_with_tools_openai.yaml"):
+        llm_with_tools.invoke("What is the sum of 1 and 2?")
+
+
+@pytest.mark.snapshot(ignores=["metrics.langchain.tokens.total_cost"])
```
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_community.py:1290]: Same as above

**Review by Yun-Kim — COMMENTED** [review]
Looks great, thanks for adding this support Sam!

File: ddtrace/contrib/langchain/patch.py (hunk 1 above)
Comment (sabrenner) [reply ddtrace/contrib/langchain/patch.py:417]: oop makes sense 😅

File: tests/contrib/langchain/test_langchain_community.py (hunk 5 above)
Comment (sabrenner) [reply tests/contrib/langchain/test_langchain_community.py:1270]: so true 🥲 will do!

**Review by Yun-Kim — APPROVED** [review]

**Pull request merged**